src/org/labkey/test/components/glassLibrary/components/MultiMenu.java

```diff
@@ -60,16 +62,67 @@ public void doMenuAction(List<String> pathToAction)
         }
     }
 
+    private List<String> waitForDataThenGetMenuText()
+    {
+        List<WebElement> topMenuList;
+        List<String> menuText = new ArrayList<>();
+
+        boolean stale;
+        boolean loading;
+        int tries = 1;
+
+        do {
+
+            stale = false;
+            loading = false;
+
+            try
+            {
+
+                topMenuList = Locator.tagWithAttribute("a", "role", "menuitem").findElements(this);
+                menuText = new ArrayList<>();
+
+                for (WebElement menuItem : topMenuList) {
+
+                    String menuItemText = menuItem.getText();
+
+                    // Check to see if we are still loading data.
+                    if(menuItemText.toLowerCase().contains("loading"))
+                        loading = true;
+
+                    menuText.add(menuItem.getText());
+
+                }
+
+            }
+            catch(StaleElementReferenceException staleExc)
+            {
+                // This happens in the time between the change of the menu content from containing "loading"
+                // to having data (like "sample sets").
+                stale = true;
+            }
+
+            // Just a small pause, no need to keep hitting the DOM if it looks like the server is doing something
+            if(stale || loading )
+                WebDriverWrapper.sleep(500);
+
+            tries++;
+
+            // Keep trying until we get valid menu text.
+        } while((stale || loading) && (tries <= 10));
+
+        if(tries > 10)
+            throw new RuntimeException("Menu text never returned.");
```

Comment:
Use the stale and loading flags to trigger this error state. It's less prone to off-by-one errors and we could generate more informative error messages (maybe include the menu text if we found any before running out of time). e.g.
```
if (stale)
{
    throw new RuntimeException("Menu items kept going stale");
}
if (loading)
{
    throw new RuntimeException("Menu items still loading. " + menuText.toString());
}
```

Reply:
Updated.